Realm Pick Algorithm v2 (#145)
decentraland/adr: content/ADR-145-pick-realm-v2.md

- The max amount of users for the server is reached
- The CPU is at 90% or higher
- The memory usage is at 90% or higher

_Pseudo Code_:

To filter out a new check will need to be added to:
https://github.com/decentraland/kernel/blob/main/packages/shared/dao/index.ts#L69

```typescript
export function fetchCatalystStatus() {

  if (currentChecks && result.accepting_users) {
    // ...
  }

  return undefined
}
```

_Configuration_:

This configuration will be at BFF (Catalyst) level

```typescript
export type AcceptingUsersConfig = {
  config?: 
          { maxAmountOfUsers: number;
            maxCpu: number;
            maxMemory: number;
          }
}
```

### Catalyst Version Link

_Description_:

Each Catalyst (BFF) exposes the version of Content and Lambdas that it's using. The Kernel should read that field, compare them and prioritize the ones having the highest value.
As Content Server and Lambdas share the same version for each Catalyst and they are using semver, there is always an unique way to sort the candidates.

_Pseudo Code_:

```typescript
export function versionLink() {

>   const catalystsWithVersio: { realm: string, version: string } = peers.sortByVersion()

Reviewer: ```suggestion
  const catalystsWithVersion: { realm: string, version: string } = peers.sortByVersion()
```

  const versionBuckets: PriorityQueue<{ realm: string, version: string }[]> = groupByVersions(catalystsWithVersion)
  // This selects randomly between the candidates in the selected bucket 

>   // And choses the bucket with the same algorithm used for close peers

Reviewer: ```suggestion
  // And chooses the bucket with the same algorithm used for close peers
```

---
layout: adr
adr: 145
title: Pick Realm Algorithm V2
date: 2022-11-28
status: Draft
type: RFC
spdx-license: CC0-1.0
slug: /adr/TEMPLATE
authors:
  - agusaldasoro
---

## Abstract

> The logic on the realm scoring algorithm should be improved to take into consideration more variables in order for a server to be more eligible. As with the current model, any added variable needs to be modifiable through Feature Flags so that they can be changed in case of need without a release or a reboot of the server.

Reviewer 2: The abstract does not describe the document's current status nor summarize its content.

## Context, Reach & Prioritization

Some aspects that should be considered:

- Resources: if a server is with the CPU at 90%, don't add more load to that server, same for memory
- Catalyst version: if the version of the node is behind the network, it should be less eligible

## Solution Space Exploration

> The current implementation as described in [Realm Picking Algorithm ADR](/adr/ADR-86) is expansible by adding new links. In this proposal two new links need to be added: resources and version, as the traffic will need to be prioritized to instances that have the latest version and more resources available.

Reviewer 2: I cannot use the link `Realm Picking Algorithm ADR` (I tried to open it [here](https://github.com/decentraland/adr/blob/c6c8b45c01a67e3afb99ff216c55f8ca4c9ca2fa/content/ADR-145-pick-realm-v2.md)). Should we use an absolute link instead?

Reviewer: @kevinszuchet This link works ok when you use the rendered version (you can get the URL from the PR's preview link: https://51d95b33.adr-cvq.pages.dev/adr/ADR-145).

Author: I copied the format from ADR-1

## Specification

### Resources Link

> Expose publicly the resources availables of a Catalyst Node is a security risk, so the estimation of resources available should be done in the Catalyst (specifically in the BFF) and not in Kernel.

Reviewer 2: What do you think about this suggestion?

```suggestion
Publicly exposing the Catalyst Node's available resources is a security risk, so their estimation should be done in the Catalyst (specifically in the BFF) and not in the Kernel.
```

Reviewer 2: Would you mind clarifying what's the meaning of `BFF`? 🙏

> This way the BFF will levearege the usage of the field `accepting_users` in `/about` endpoint. 

Reviewer 2: Small suggestion:
```suggestion
This way, the BFF will leverage the usage of the field `accepting_users` in the `/about` endpoint.
```

> The logic from the Kernel side will be very simple: it will filter out the catalysts that have that value in false.

Reviewer 2: Small suggestion:
```suggestion
The logic from the Kernel side will be simple: it will filter out the catalysts having that value as false.
```

> From the other hand, the BFF will have a set of rules to calculate that value. The accepting users field will be false if:

Reviewer 2: Small suggestion:
```suggestion
On the other hand, the BFF will have a set of rules to calculate that value. The accepting users' field will be false if:
```

- The max amount of users for the server is reached
- The CPU is at 90% or higher
- The memory usage is at 90% or higher

### Catalyst Version Link

> Each Catalyst (BFF) exposes the version of Content and Lambdas that it's using. Kernel should read that field, compare them and prioritize the ones that have the greatest value.

Reviewer 2: Small suggestion:
```suggestion
Each Catalyst (BFF) exposes the version of Content and Lambdas that it's using. The Kernel should read that field, compare them and prioritize the ones having the highest value.
```

---
layout: adr
adr: 145
title: Pick Realm Algorithm V2
date: 2022-11-28
status: Draft
type: RFC
spdx-license: CC0-1.0
slug: /adr/TEMPLATE
authors:
  - agusaldasoro
---

## Abstract

This document describes the changes needed in the realm picking algorithm to take into account the resources of the computer and the catalyst version.

The logic on the realm scoring algorithm should be improved to take into consideration more variables in order for a server to be more eligible. As with the current model, any added variable needs to be modifiable through Feature Flags so that they can be changed in case of need without a release or a reboot of the server.

## Context, Reach & Prioritization

Some aspects that should be considered:

> - Resources: if a server is with the CPU at 90%, don't add more load to that server, same for memory

Reviewer: ```suggestion
- Resources: if a server has either CPU or memory at 90%, don't add more load to that server.
```

> - Catalyst version: if the version of the node is behind the network, it should be less eligible

Reviewer: ```suggestion
- Catalyst version: if the version of the node is behind the network, it should be less eligible.
```

## Solution Space Exploration

> The current implementation as described in [Realm Picking Algorithm ADR](/adr/ADR-86) is expansible by adding new links. In this proposal two new links need to be added: resources and version, as the traffic will need to be prioritized to instances that have the latest version and more resources available.

Reviewer: ```suggestion
The current implementation (as described in [Realm Picking Algorithm ADR](/adr/ADR-86)) is expansible by adding new links. In this proposal two new links will be added: resources and version, so that the algorithm can prioritize sending traffic to instances that have the latest version and more resources available.
```

## Specification

### Resources Link

Publicly exposing the Catalyst Node's available resources is a security risk, so their estimation should be done in the Catalyst (specifically in the BFF) and not in the Kernel.

This way, the BFF will leverage the usage of the field `accepting_users` in the `/about` endpoint.

The logic from the Kernel side will be simple: it will filter out the catalysts having that value as false.

> On the other hand, the BFF will have a set of rules to calculate that value. The accepting users' field will be false if:

Reviewer: ```suggestion
On the other hand, the BFF will have a set of rules to calculate that value. The `accepting_users` field will be false if:
```
